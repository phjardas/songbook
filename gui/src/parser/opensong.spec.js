import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import parse from './opensong';

const readFile = promisify(fs.readFile);

async function parseSong(name) {
  const file = path.resolve(__dirname, `opensong.spec.${name}.txt`);
  const content = await readFile(file, 'utf8');
  return parse(content);
}

describe('parser', () => {
  describe('opensong', () => {
    it('should parse a song', async () => {
      const { messages, sections } = await parseSong('song1');

      expect(messages).toEqual([]);
      expect(sections).toHaveLength(5);

      expect(sections[0]).toHaveProperty('type', undefined);
      expect(sections[0]).toHaveProperty('label', undefined);
      expect(sections[0].lines).toEqual([
        [
          { text: 'I ', col: 0 },
          { text: 'looked over Jordan, and ', chord: 'D', col: 2 },
          { text: 'what did I ', chord: 'G', col: 26 },
          { text: 'see,', chord: 'D', col: 37 },
        ],
        [{ text: 'Comin’ for to carry me ', col: 0 }, { text: 'home.', chord: 'A7', col: 23 }],
      ]);

      expect(sections[1]).toHaveProperty('type', 'chorus');
      expect(sections[1]).toHaveProperty('label', 'Chorus');

      expect(sections[2]).toHaveProperty('type', 'verse');
      expect(sections[2]).toHaveProperty('label', 'Verse 2');

      expect(sections[3]).toHaveProperty('type', 'bridge');
      expect(sections[3]).toHaveProperty('label', 'Bridge');
      expect(sections[3].lines).toEqual([
        [{ chord: 'D', col: 0 }, { chord: 'G', col: 2 }, { chord: 'D', col: 4 }],
        [{ chord: 'A7', col: 0 }],
      ]);

      expect(sections[4]).toHaveProperty('type', 'verse');
      expect(sections[4]).toHaveProperty('label', 'Verse 3');
      expect(sections[4].lines).toEqual([
        [{ text: 'I looked over Jordan, and what did I see,', col: 0 }],
        [{ text: 'Comin’ for to carry me home.', col: 0 }],
      ]);
    });
  });
});
