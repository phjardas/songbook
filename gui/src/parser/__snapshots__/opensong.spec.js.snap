// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser opensong should parse a song 1`] = `
Object {
  "messages": Array [],
  "sections": Array [
    Object {
      "firstLine": 0,
      "label": undefined,
      "lastLine": 3,
      "lines": Array [
        Object {
          "lines": Array [
            0,
            1,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "I ",
            },
            Object {
              "chord": "D",
              "col": 3,
              "text": "looked over Jordan, and ",
            },
            Object {
              "chord": "G",
              "col": 27,
              "text": "what did I ",
            },
            Object {
              "chord": "D",
              "col": 38,
              "text": "see,",
            },
          ],
        },
        Object {
          "lines": Array [
            2,
            3,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "Comin’ for to carry me ",
            },
            Object {
              "chord": "A7",
              "col": 24,
              "text": "home.",
            },
          ],
        },
      ],
      "type": undefined,
    },
    Object {
      "firstLine": 5,
      "label": "Chorus",
      "lastLine": 9,
      "lines": Array [
        Object {
          "lines": Array [
            6,
            7,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "Swing ",
            },
            Object {
              "chord": "D",
              "col": 7,
              "text": "low, sweet ",
            },
            Object {
              "chord": "G",
              "col": 18,
              "text": "chari",
            },
            Object {
              "chord": "D",
              "col": 23,
              "text": "ot,",
            },
          ],
        },
        Object {
          "lines": Array [
            8,
            9,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "Comin’ for to carry me ",
            },
            Object {
              "chord": "A7",
              "col": 24,
              "text": "home.",
            },
          ],
        },
      ],
      "type": "chorus",
    },
    Object {
      "firstLine": 11,
      "label": "Verse 2",
      "lastLine": 15,
      "lines": Array [
        Object {
          "lines": Array [
            12,
            13,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "I ",
            },
            Object {
              "chord": "D",
              "col": 3,
              "text": "looked over Jordan, and ",
            },
            Object {
              "chord": "G",
              "col": 27,
              "text": "what did I ",
            },
            Object {
              "chord": "D",
              "col": 38,
              "text": "see,",
            },
          ],
        },
        Object {
          "lines": Array [
            14,
            15,
          ],
          "parts": Array [
            Object {
              "chord": undefined,
              "col": 1,
              "text": "Comin’ for to carry me ",
            },
            Object {
              "chord": "A7",
              "col": 24,
              "text": "home.",
            },
          ],
        },
      ],
      "type": "verse",
    },
    Object {
      "firstLine": 17,
      "label": "Bridge",
      "lastLine": 19,
      "lines": Array [
        Object {
          "lines": Array [
            18,
          ],
          "parts": Array [
            Object {
              "chord": "D",
              "col": 1,
            },
            Object {
              "chord": "G",
              "col": 3,
            },
            Object {
              "chord": "D",
              "col": 5,
            },
          ],
        },
        Object {
          "lines": Array [
            19,
          ],
          "parts": Array [
            Object {
              "chord": "A7",
              "col": 1,
            },
          ],
        },
      ],
      "type": "bridge",
    },
    Object {
      "firstLine": 21,
      "label": "Verse 3",
      "lastLine": 23,
      "lines": Array [
        Object {
          "lines": Array [
            22,
          ],
          "parts": Array [
            Object {
              "col": 1,
              "text": "I looked over Jordan, and what did I see,",
            },
          ],
        },
        Object {
          "lines": Array [
            23,
          ],
          "parts": Array [
            Object {
              "col": 1,
              "text": "Comin’ for to carry me home.",
            },
          ],
        },
      ],
      "type": "verse",
    },
  ],
}
`;
